app/screens/edit_server/form.tsx

```diff
@@ -0,0 +1,194 @@
+    formContainer: {
+        alignItems: 'center',
+        maxWidth: 600,
+        width: '100%',
+        paddingHorizontal: 20,
+    },
+    fullWidth: {
+        width: '100%',
+    },
+    error: {
+        marginBottom: 18,
+    },
+    chooseText: {
+        alignSelf: 'flex-start',
+        color: changeOpacity(theme.centerChannelColor, 0.64),
+        marginTop: 8,
+        ...typography('Body', 75, 'Regular'),
+    },
+    connectButton: {
+        backgroundColor: changeOpacity(theme.centerChannelColor, 0.08),
+        width: '100%',
+        marginTop: 32,
+        marginLeft: 20,
+        marginRight: 20,
+        padding: 15,
+    },
+    loadingContainerStyle: {
+        marginRight: 10,
+        padding: 0,
+        top: -2,
+    },
+    loading: {
+        height: 20,
+        width: 20,
+    },
+}));
+
+const EditServerForm = ({
+    buttonDisabled,
+    connecting,
+    displayName = '',
+    displayNameError,
+    handleUpdate,
+    handleDisplayNameTextChanged,
+    keyboardAwareRef,
+    serverUrl,
+    theme,
+}: Props) => {
+    const {formatMessage} = useIntl();
+    const isTablet = useIsTablet();
+    const dimensions = useWindowDimensions();
+    const displayNameRef = useRef<FloatingTextInputRef>(null);
+    const styles = getStyleSheet(theme);
+
+    const onBlur = useCallback(() => {
+        if (Platform.OS === 'ios') {
+            const reset = !displayNameRef.current?.isFocused();
+            if (reset) {
+                keyboardAwareRef.current?.scrollToPosition(0, 0);
+            }
+        }
+    }, []);
+
+    const onUpdate = useCallback(() => {
+        Keyboard.dismiss();
+        handleUpdate();
+    }, [handleUpdate]);
+
+    const onFocus = useCallback(() => {
+        if (Platform.OS === 'ios') {
+            let offsetY = 160;
+            if (isTablet) {
+                const {width, height} = dimensions;
+                const isLandscape = width > height;
+                offsetY = isLandscape ? 230 : 100;
+            }
+            requestAnimationFrame(() => {
+                keyboardAwareRef.current?.scrollToPosition(0, offsetY);
+            });
+        }
+    }, [dimensions, isTablet]);
+
+    useEffect(() => {
+        if (Platform.OS === 'ios' && isTablet) {
+            if (displayNameRef.current?.isFocused()) {
+                onFocus();
+            } else {
+                keyboardAwareRef.current?.scrollToPosition(0, 0);
+            }
+        }
+    }, [onFocus]);
+
+    const buttonType = buttonDisabled ? 'disabled' : 'default';
+    const styleButtonText = buttonTextStyle(theme, 'lg', 'primary', buttonType);
+    const styleButtonBackground = buttonBackgroundStyle(theme, 'lg', 'primary', buttonType);
+
+    let buttonID = t('edit_server.save');
+    let buttonText = 'Save';
+    let buttonIcon;
+
+    if (connecting) {
+        buttonID = t('edit_server.saving');
+        buttonText = 'Saving';
+        buttonIcon = (
```

Comment:
For my own understanding, upon re-rendering and while connecting is true, would `buttonIcon` this form create a new instance of the `Loading` component ? In that case, we can `useMemo` it, but still I guess the impact is near zero and hence can be avoided.

Reply:
Why would I over optimize if there is no need?